src/hooks/useActiveDescendantListFocus/index.js

```diff
@@ -0,0 +1,157 @@
+
+function useActiveDescendantListFocus({
+  focusedElementRef, // the reference for the component that listens to keyboard
+  itemsIds,
+  isItemSelectable,
+  onItemClick,
+  focusedElementRole = ROLES.GROUP,
+  isHorizontalList = false,
+  useDocumentEventListeners = false
+}) {
+  const itemsCount = itemsIds.length;
+  const nextArrow = isHorizontalList ? ARROW_DIRECTIONS.RIGHT : ARROW_DIRECTIONS.DOWN;
+  const backArrow = isHorizontalList ? ARROW_DIRECTIONS.LEFT : ARROW_DIRECTIONS.UP;
+
+  const [visualFocusItemIndex, setVisualFocusItemIndex] = useState(-1);
+  const triggerByKeyboard = useRef(false);
+  const onArrowKeyEvent = useCallback(
+    direction => {
+      if (document.activeElement === focusedElementRef.current) {
+        let newIndex;
+        triggerByKeyboard.current = true;
+
+        if (direction === nextArrow) {
+          if (visualFocusItemIndex > itemsCount - 1) return;
+          // Go over all the next items until found one which is selectable
+          for (let offset = 1; offset <= itemsCount; offset++) {
+            const nextIndex = (visualFocusItemIndex + offset) % itemsCount;
+            if (isItemSelectable(nextIndex)) {
+              newIndex = nextIndex;
+              break;
+            }
+          }
+        } else if (direction === backArrow) {
+          for (let offset = 1; offset <= itemsCount - 1; offset++) {
+            let prevIndex = visualFocusItemIndex - offset;
+            if (prevIndex < 0) {
+              prevIndex = itemsCount + prevIndex;
+            }
+            if (isItemSelectable(prevIndex)) {
+              newIndex = prevIndex;
+              break;
+            }
+          }
+        }
+
+        if (newIndex > -1 && newIndex !== visualFocusItemIndex) setVisualFocusItemIndex(newIndex);
+      }
+    },
+    [focusedElementRef, nextArrow, backArrow, visualFocusItemIndex, itemsCount, isItemSelectable]
+  );
+  const onArrowBack = useCallback(() => {
+    onArrowKeyEvent(backArrow);
+  }, [backArrow, onArrowKeyEvent]);
+
+  const onArrowNext = useCallback(() => {
+    onArrowKeyEvent(nextArrow);
+  }, [nextArrow, onArrowKeyEvent]);
+
+  const baseOnClickCallback = useCallback(
+    (event, itemIndex) => {
+      const hasValidIndex = itemIndex >= 0 && itemIndex < itemsCount;
+      if (!onItemClick || !hasValidIndex || !isItemSelectable(itemIndex)) return;
+      if (visualFocusItemIndex !== itemIndex) setVisualFocusItemIndex(itemIndex);
+      onItemClick(event, itemIndex);
+      triggerByKeyboard.current = event instanceof MouseEvent;
+      // If click is triggered by keyboard return focus on the input again
+      if (!triggerByKeyboard.current) {
+        requestAnimationFrame(() => focusedElementRef.current?.focus());
+      }
+    },
+    [itemsCount, onItemClick, isItemSelectable, visualFocusItemIndex, focusedElementRef]
+  );
+
+  const keyboardOnSelectCallback = useCallback(
+    event => {
+      if (focusedElementRef.current.contains(document.activeElement)) {
```

Comment:
I'm not 100% sure I understand this part - perhaps it's best adding some comment?

Reply:
I want to trigger the visual focus on any list option only if the element which attached to the component really have natural focus right now. I will add a comment :)